standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1236,31 +1237,60 @@ public void replTableWriteIdState(ReplTblWriteIdStateRequest rqst) throws MetaEx
     return abortedWriteIds;
   }
 
+  private ValidTxnList getValidTxnList(Connection dbConn, String fullTableName, Long writeId) throws MetaException,
+          SQLException {
+    PreparedStatement pst = null;
+    ResultSet rs = null;
+    try {
+      String[] names = TxnUtils.getDbTableName(fullTableName);
+      String s = "select t2w_txnid from TXN_TO_WRITE_ID where  t2w_database = ? and t2w_table = ? and t2w_writeid = ?";
+      pst = dbConn.prepareStatement(sqlGenerator.addEscapeCharacters(s));
+      pst.setString(1, names[0]);
+      pst.setString(2, names[1]);
+      pst.setLong(3, writeId);
+      LOG.debug("Going to execute query <" + s.replaceAll("\\?", "{}") + ">", quoteString(names[0]),
+              quoteString(names[1]), writeId);
+      rs = pst.executeQuery();
+      if (rs.next()) {
+        return TxnCommonUtils.createValidReadTxnList(getOpenTxns(), rs.getLong(1));
+      }
+      throw new MetaException("invalid write id " + writeId + " for table " + fullTableName);
+    } finally {
+      close(rs);
+      if (pst != null) {
+        pst.close();
+      }
+    }
+  }
+
   @Override
   @RetrySemantics.ReadOnly
-  public GetValidWriteIdsResponse getValidWriteIds(GetValidWriteIdsRequest rqst)
-          throws NoSuchTxnException, MetaException {
+  public GetValidWriteIdsResponse getValidWriteIds(GetValidWriteIdsRequest rqst) throws MetaException {
     try {
       Connection dbConn = null;
       Statement stmt = null;
       ValidTxnList validTxnList;
 
-      // We should prepare the valid write ids list based on validTxnList of current txn.
-      // If no txn exists in the caller, then they would pass null for validTxnList and so it is
-      // required to get the current state of txns to make validTxnList
-      if (rqst.isSetValidTxnList()) {
-        validTxnList = new ValidReadTxnList(rqst.getValidTxnList());
-      } else {
-        // Passing 0 for currentTxn means, this validTxnList is not wrt to any txn
-        validTxnList = TxnCommonUtils.createValidReadTxnList(getOpenTxns(), 0);
-      }
       try {
         /**
          * This runs at READ_COMMITTED for exactly the same reason as {@link #getOpenTxnsInfo()}
          */
         dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
         stmt = dbConn.createStatement();
 
+        // We should prepare the valid write ids list based on validTxnList of current txn.
+        // If no txn exists in the caller, then they would pass null for validTxnList and so it is
+        // required to get the current state of txns to make validTxnList
+        if (rqst.isSetValidTxnList()) {
+          assert rqst.isSetWriteId() == false;
+          validTxnList = new ValidReadTxnList(rqst.getValidTxnList());
+        } else if (rqst.isSetWriteId()) {
+          validTxnList = getValidTxnList(dbConn, rqst.getFullTableNames().get(0), rqst.getWriteId());
```

Comment:
What is the behaviour if write ID is set and is 0? It can happen if source alter doesn't happen within write id.

Reply:
write id is set only at target side and only for transactional table where write id must be there ..if write id is set to 0 by mistake ..it will throw invalid write id exception